src/ClientTransactionTask.h

```diff
@@ -0,0 +1,264 @@
+        /// Cached object value.  Used to service reads and store values for
+        /// committing writes.  Ideally this would be a unique pointer to
+        /// manage the memory automatically but std::multimap is missing the
+        /// emplace feature.
+        ObjectBuffer* objectBuf;
+        /// Conditions upon which the transaction operation associated with
+        /// this object should abort.
+        RejectRules rejectRules;
+
+        /// The rpcId to uniquely identify this operation.
+        uint64_t rpcId;
+        /// Used to keep track of what stage in the commit process this
+        /// operation has reached.
+        enum { PENDING, PREPARE, DECIDE, FAILED } state;
+
+        /// Default constructor for CacheEntry.
+        CacheEntry()
+            : type(CacheEntry::INVALID)
+            , objectBuf(NULL)
+            , rejectRules({0, 0, 0, 0, 0})
+            , rpcId(0)
+            , state(PENDING)
+        {}
+
+        /// Copy constructor for CacheEntry, used to get around the missing
+        /// emplace feature in std::multimap.
+        explicit CacheEntry(const CacheEntry& other)
+            : type(other.type)
+            , objectBuf(other.objectBuf)
+            , rejectRules(other.rejectRules)
+            , rpcId(other.rpcId)
+            , state(other.state)
+        {}
+
+        /// Destructor for CacheEntry.
+        ///
+        /// Warning: Multiple copies of CacheEntry objects may cause the
+        /// ObjectBuffer pointed in the entry to be double freed.  This
+        /// is indirectly due to missing emplace feature in std::multimap.
+        ~CacheEntry()
+        {
+            if (objectBuf)
+                delete objectBuf;
+        }
+
+        /// Assignment operator for CacheEntry, used to get around the missing
+        /// emplace feature in std::multimap.
+        CacheEntry& operator=(const CacheEntry& other)
+        {
+            if (this != &other) {
+                type = other.type;
+                objectBuf = other.objectBuf;
+                rejectRules = other.rejectRules;
+                rpcId = other.rpcId;
+                state = other.state;
+            }
+            return *this;
+        }
+    };
+
+    explicit ClientTransactionTask(RamCloud* ramcloud);
+
+    CacheEntry* findCacheEntry(Key& key);
+    WireFormat::TxDecision::Decision getDecision();
+    Status getStatus() { return status; }
+    CacheEntry* insertCacheEntry(uint64_t tableId, const void* key,
+            uint16_t keyLength, const void* buf, uint32_t length);
+    /// Check if the task as completed the commit protocol.
+    bool isReady() { return (state == DONE); }
+    void performTask();
+
+  PRIVATE:
+    // Forward declaration of RPCs
+    class PrepareRpc;
+    class DecisionRpc;
+
+    /// Overall client state information.
+    RamCloud* ramcloud;
+
+    /// Number of participant objects/operations.
+    uint32_t participantCount;
+    /// List of participant object identifiers.
```

Comment:
Let's discuss.  The idea was making RPC sterilization more efficient.  Maybe there is a better way.